hw/ip/pattgen/doc/_index.md

```diff
@@ -0,0 +1,137 @@
+
+# Overview
+
+This document specifies PATTGEN hardware IP (HWIP) functionality.
+This module conforms to the [Comportable guideline for peripheral functionality.]({{< relref "doc/rm/comportability_specification/index.md" >}})
+See that document for integration overview within the broader top-level system.
+
+## Features
+
+- Generates time-dependent patterns on two (2) channels, each with its own clock.
+   - In each channel, data is transmitted serially on a one-bit data (`pda`) output, which is synchronous to a corresponding parallel clock signal (`pcl`).
+   - The channels may operate independently or synchronously with each other.
+- Each output channel supports the following configuration registers:
+    - Pattern data per output (up to 64 bits of data).
+    - Pattern length per output (between 1 and 64 bit).
+    - 32-bit pre-divider to derive pattern clock from I/O clock (minimum ratio: 2).
+    - A 10-bit repeat counter
+    - A bit to control the polarity of the clock signal
+- Sends interrupt on pattern completion.
+## Description
+
+The PATTGEN IP transmits short (maximum 64 bits) time-dependent data patterns on two clock-parallel channels.
+Each channel consists of one clock (`pcl`) and one data (`pda`) line.
+The output channels may be activated and operated independently, or they can be started at the same time to effectively create a 4-output pattern.
+
+## Compatibility
+
+This IP block does not have any direct hardware compatibility requirements.
+
+# Theory of Operations
+
+The pattern can be started (or halted) on either channel by setting the corresponding `CTRL.ENABLE` bit to 1 (or 0) for the desired channel.
+Once disabled, either channel can be configured independently so long as the desired channel is first disabled.
+The channel parameters (i.e. clock divider ratio, clock polarity, pattern length, pattern data, and repetition count) can all be programmed on a per-channel basis.
+Enabling the pattern generator channel starts the pattern from the beginning.
+
+Please note that writes to a channel's configuration registers have no effect while the channel is enabled.
+For operational simplicity, the configuration registers are only transferred into the internal finite state machines while a channel is disabled.
+Changes to the configuration registers only take effect once the channel has been disabled and re-enabled.
+
+## Block Diagram
+
+![](pattgen_block_diagram.svg)
+
+## Hardware Interfaces
+
+{{< hwcfg "hw/ip/pattgen/data/pattgen.hjson" >}}
+
+## Design Details
+
+The design consists of two identical and independent finite state machines, each an instance of module `pattgen_fsm`.
+Each FSM is essentially three nested counters, with one counter to control the clock division, another to count out the sequence bits, and a third to keep count of the number of repetitions.
+
+Each FSM consists of.
+- Inputs:
+    - `clk_io`, `reset`, `enable`, `clk_pre_divide`, `pattern`, `pattern_size`, `polarity`,  and `n_repeats`
```

Comment:
Clearly not for this PR, but you might want to align the names of the registers with the names of the internal signals in a follow-up (either by renaming registers, or by renaming internal signals) (e.g. pattern -> data, pattern_size -> len, n_repeats -> repeats).

Reply:
Good point.  I'll be putting together a draft of the IP today.